internal/servers/controller/controller.go

```diff
@@ -124,6 +123,24 @@ func New(ctx context.Context, conf *Config) (*Controller, error) {
 		}
 	}
 
+	for _, l := range conf.Listeners {
+		if len(l.Config.Purpose) != 1 {
+			return nil, fmt.Errorf("listener must have exactly one purpose")
+		}
+		switch l.Config.Purpose[0] {
+		case "api":
+			c.apiListeners = append(c.apiListeners, l)
+		case "cluster":
+			c.clusterListeners = append(c.clusterListeners, l)
+		}
+	}
+	if len(c.apiListeners) == 0 {
+		return nil, fmt.Errorf("no api listeners found")
+	}
+	if len(c.clusterListeners) == 0 {
+		return nil, fmt.Errorf("no cluster listeners found")
+	}
```

Comment:
should this check for one and only one?

Reply:
Yeah, this is still outstanding and linked to https://github.com/hashicorp/boundary/pull/1874#pullrequestreview-899415126. Currently waiting on someone to shed light on whether we're supposed to have multiple clusters or not - perhaps I should assume in this PR that it'd just ever be one cluster and go from there?

In which case, `clusterListeners` really becomes `clusterListener`

Reply:
https://github.com/hashicorp/boundary/pull/1874#discussion_r819602594